### PR TITLE
rocksdb/util: Fixes for http://tracker.ceph.com/issues/20529 

### PR DESCRIPTION
**Pull request by oliveiradan**
This PR goes along with: https://github.com/ceph/ceph/pull/17347

### CODE REVIEW

File: util/crc32c.cc [hunk 1]
```diff
@@ -286,12 +285,141 @@ static const uint32_t table3_[256] = {
...
+  SSE_NONE,
+};
+
+#if !defined(__has_builtin)
+  #define __has_builtin(x) false
+#endif
+
+// We want to know if a cpu feature is available at runtime. 
+static inline bool is_cpu_feature_available(const CpuSSEFeature& cpu_feature) {
+#if (defined(__GNUC__) || (__clang__))
```
Comment (tchaikov) [inline util/crc32c.cc]: clang defines `__GNUC__`.

File: util/crc32c.cc [hunk 2]
```diff
@@ -404,5 +542,12 @@ uint32_t Extend(uint32_t crc, const char* buf, size_t size) {
   return ChosenExtend(crc, buf, size);
 }
 
+// We want to check if this cpu feature is available at runtime, 
+// *not at compile time*. 
+static inline bool isRuntimeSSE42Available() {
```
Comment (tchaikov) [inline util/crc32c.cc]: this function is not used.

File: util/crc32c.cc [hunk 3]
```diff
@@ -1,7 +1,7 @@
 //  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
-//  This source code is licensed under both the GPLv2 (found in the
-//  COPYING file in the root directory) and Apache 2.0 License
-//  (found in the LICENSE.Apache file in the root directory).
+//  This source code is licensed under the BSD-style license found in the
+//  LICENSE file in the root directory of this source tree. An additional grant
+//  of patent rights can be found in the PATENTS file in the same directory.
```
Comment (tchaikov) [inline util/crc32c.cc]: why change the license? i think only the copyright holder(s) is allowed to do so.

Comment (oliveiradan) [reply util/crc32c.cc]: I didn't really changed it, but only kept the same as what we have on the rocksdb commit used by Ceph, which is not the rocksdb:master by the way.

File: util/crc32c.cc (hunk 2 above)
Comment (oliveiradan) [reply util/crc32c.cc]: Well, that's a functionality added. We can used it to check for a specific feature at runtime. That's why it is ther.e

File: util/crc32c.cc (hunk 1 above)
Comment (oliveiradan) [reply util/crc32c.cc]: True. I spaced it when putting in the code. I will fix it.

**Comment by alfredodeza** [comment]
You are submitting a PR against a fork of rocksdb, please submit them first to the origin repo, and once that gets merged it can then be added here (will defer on closing this, since @tchaikov is reviewing)

**Comment by oliveiradan** [comment]
@alfredodeza , 

I am sorry, but as this was my first time submitting anything to a **submodule (rocksdb) for Ceph**, my understanding is that I had to submit it to *origin repo* which is my fork of https://github.com/ceph/rocksdb . Am I missing something?

**Comment by alfredodeza** [comment]
@oliveiradan this is a fork from https://github.com/facebook/rocksdb so we prefer changes to be merged there ("upstream") so that they can be used with this fork

**Comment by oliveiradan** [comment]
@alfredodeza,

When I started working on this, I asked **Sage** *and *my understanding was that we wanted to work based on our (ceph) fork of the rocksdb project* (https://github.com/ceph/rocksdb), most likely due the fact we are not using the latest master commit?! So, not sure that even if we push it out there ("upstream"), that: 
1. It will get merged any soon. 
2. If/after it gets merged, than we (ceph) would need to get that update only? Once we are not running the latest version/commit from  https://github.com/facebook/rocksdb 

So, please how exactly should we proceed here? what's the proper way to submit a changes to submodules? Always submit them "upstream" first? 

Thanks!

**Comment by tchaikov** [comment]
please note, upstream has merged https://github.com/facebook/rocksdb/pull/2807, and the change on ceph side is posted at https://github.com/ceph/ceph/pull/17388.

**Comment by tchaikov** [comment]
closing in favor of  facebook#2807

**Pull request closed without merging**